File: linked_lists/sum_lists/sol1.py
import sys

sys.path.append('../')

from sll import sll

def sum_lists(sll1, sll2):
  nd1 = sll1.head
  nd2 = sll2.head
  
  sll3 = sll.sll()
  
  add_one = 0
  key = 0
  while nd1 and nd2:
    data = nd1.data + nd2.data + add_one
    add_one = data // 10
    data = data % 10
    sll3.addLast(key, data)
    key += 1
    nd1 = nd1.next
    nd2 = nd2.next
    
  nd = None
  if nd1:
    nd = nd1
  elif nd2:
    nd = nd2
    
  while nd:
    data = nd.data + add_one
    add_one = data // 10
    data %= 10
    sll3.add_Last(key, data)
    key ++ 1
    nd = nd.next
    
  return sll3
  
if __name__ == "__main__":
  slinked_list1 = sll.sll()
  key = 0
  value = [7, 1, 6]
  for it in value:
    slinked_list1.addLast(key, it)
    key += 1
  slinked_list2 = sll.sll()
  key = 0
  value = [5, 9, 2]
  for it in value:
    slinked_list2.addLast(key, it)
    key += 1
    
  print slinked_list1
  print slinked_list2
  
  print "add"
  
  print sum_lists(slinked_list1, slinked_list2)

File: linked_lists/sum_lists/sol1.cpp
#include "sll.h"

#include <iostream>
#include <vector>

node* sumLists(node* nd1, node* nd2) {
   unsigned int add = 0;
   int k = 0;
   node* head = 0;
   node* tail = 0;
   while ( nd1 || nd2 ) {
      int s = 0;
      if ( nd1 ) {
         s += nd1->data();
         nd1 = nd1->next();
      }
      if ( nd2 ) {
         s += nd2->data();
         nd2 = nd2->next();
      }
      s += add;
      node* nd = new node(s%10, k);
      if ( !head ) {
         head = nd;
         tail = head;
      } else {
         tail->setNext(nd);
         tail = nd;
      }
      add = s / 10;
      ++k;
   }
   if ( add == 1 ) {
      node* nd = new node(add, k);
      tail->setNext(nd);
   }
   return head;
}


int main() {
  sll slinked_list;
  sll slinked_list1;
  std::vector<int> value{1, 2, 3};
  std::vector<int> value1{9, 2, 1};
  int key = 0;
  for (auto& it : value) {
    std::cout << key << std::endl;
    if (key == 0) {
      slinked_list.insertHead(it, key);
    } else {
      slinked_list.insertLast(it, key);
    }
    ++key;
  }
  
  for (auto& it : value1) {
    std::cout << key << std::endl;
    if (key == 0) {
      slinked_list1.insertHead(it, key);
    } else {
      slinked_list1.insertLast(it, key);
    }
    ++key;
  }
  
  node* nd = slinked_list.head();
  while (nd) {
    std::cout << nd->data() << " ";
    nd = nd->next();
  }
  std::cout << std::endl;
   
  nd = slinked_list1.head();
  while (nd) {
    std::cout << nd->data() << " ";
    nd = nd->next();
  }
  std::cout << std::endl;
   
  nd = slinked_list.head();
  node* nd1 = slinked_list1.head();
  nd = sumLists(nd, nd1);
  
  std::cout << "getting result: ";
  while (nd) {
    std::cout << nd->data() << " ";
    nd = nd->next();
  }
  std::cout << std::endl;
}
       